### ISSUE TITLE
Consider relaxing NodeId: Default bound

### ISSUE THREAD

**Issue by xDarksome** [databendlabs/openraft#890]
**Is your feature request related to a problem? Please describe.**

We're currently integrating `libp2p` with `openraft` and having a small issue of `libp2p::PeerId` not implementing `Default` (and conceptually I don't think it should).

**Describe the solution you'd like**

It would be great if you could remove the bound

**Describe alternatives you've considered**

For now, we're going to use a newtype on top of `libp2p::PeerId` or some intermediary type.

**Additional context**

It seems that the `Default` is being used quite extensively across the codebase. However, I don't think it's actually necessary - `Option<T>` should suffice.

**Comment by drmingdrmer** [comment]
I'm afraid removing `NodeId: Default` from the current codebase may not be straightforward. Types like `LeaderId` need to offer a `zero` value for uninitialized states or testing purposes. Utilizing `Option` would introduce unnecessary error handling using `unwrap()`.

For your specific scenario, a potential and straightforward solution is to encapsulate `libp2p::PeerId` within another struct that implements `Default`:

```rust
struct P2PNodeId(libp2p::PeerId);
impl Default for P2PNodeId {
  fn default() -> Self { /* ... */ }
}
```

**Comment by xDarksome** [comment]
Yes, that's exactly how we're handing it now, it works fine, aside from some additional boilerplate of converting the types back and forth.

 If there's ever going to be a refactoring/redesign of how you handle `LeaderId` internally, please, consider this issue.

**Comment by drmingdrmer** [comment]
Glad to hear that the workaround is functioning well for you.
If there's ever an opportunity to refactor, I will definitely keep your suggestion in mind.

Thank you for raising this issue.

**Comment by drmingdrmer** [comment]
Closing as resolved. The `NodeId` trait no longer requires `Default`:

```rust
pub trait NodeId
where Self: Sized + OptionalFeatures + Eq + PartialEq + Ord + PartialOrd 
          + Debug + Display + Hash + Clone + 'static
{}
```

The remaining `C::NodeId: Default` bounds are **conditional `impl` blocks** (primarily for testing), not trait requirements. This means `libp2p::PeerId` can be used directly as a `NodeId` without a newtype wrapper - those `Default` impls simply won't exist for your type, which is fine.
